File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/subscriptions/SubscriptionUsageRequest.java
package org.hyperledger.iroha.android.subscriptions;

import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import org.hyperledger.iroha.android.client.JsonEncoder;

/** Request payload for subscription usage recording. */
public final class SubscriptionUsageRequest {

  private final String authority;
  private final String unitKey;
  private final String delta;
  private final String usageTriggerId;

  private SubscriptionUsageRequest(final Builder builder) {
    this.authority = requireNonBlank(builder.authority, "authority");
    this.unitKey = requireNonBlank(builder.unitKey, "unit_key");
    this.delta = requireNumericLiteral(builder.delta, "delta");
    this.usageTriggerId = normalizeOptional(builder.usageTriggerId);
  }

  public String authority() {
    return authority;
  }

  public String unitKey() {
    return unitKey;
  }

  public String delta() {
    return delta;
  }

  public String usageTriggerId() {
    return usageTriggerId;
  }

  public Map<String, Object> toJsonMap() {
    final Map<String, Object> json = new LinkedHashMap<>();
    json.put("authority", authority);
    json.put("unit_key", unitKey);
    json.put("delta", delta);
    if (usageTriggerId != null) {
      json.put("usage_trigger_id", usageTriggerId);
    }
    return Collections.unmodifiableMap(json);
  }

  public byte[] toJsonBytes() {
    return JsonEncoder.encode(toJsonMap()).getBytes(StandardCharsets.UTF_8);
  }

  public static Builder builder() {
    return new Builder();
  }

  public static final class Builder {
    private String authority;
    private String unitKey;
    private String delta;
    private String usageTriggerId;

    private Builder() {}

    public Builder authority(final String authority) {
      this.authority = authority;
      return this;
    }

    public Builder unitKey(final String unitKey) {
      this.unitKey = unitKey;
      return this;
    }

    public Builder delta(final String delta) {
      this.delta = delta;
      return this;
    }

    public Builder usageTriggerId(final String usageTriggerId) {
      this.usageTriggerId = usageTriggerId;
      return this;
    }

    public SubscriptionUsageRequest build() {
      return new SubscriptionUsageRequest(this);
    }
  }

  private static String requireNonBlank(final String value, final String field) {
    if (value == null || value.trim().isEmpty()) {
      throw new IllegalStateException(field + " is required");
    }
    return value.trim();
  }

  private static String requireNumericLiteral(final String value, final String field) {
    final String trimmed = requireNonBlank(value, field);
    int index = 0;
    final char first = trimmed.charAt(0);
    if (first == '+' || first == '-') {
      if (first == '-') {
        throw new IllegalArgumentException(field + " must be non-negative");
      }
      index = 1;
    }
    if (index >= trimmed.length()) {
      throw new IllegalArgumentException(field + " must be numeric");
    }
    boolean seenDot = false;
    boolean seenDigit = false;
    for (int i = index; i < trimmed.length(); i++) {
      final char ch = trimmed.charAt(i);
      if (ch == '.') {
        if (seenDot) {
          throw new IllegalArgumentException(field + " must be numeric");
        }
        seenDot = true;
        continue;
      }
      if (ch < '0' || ch > '9') {
        throw new IllegalArgumentException(field + " must be numeric");
      }
      seenDigit = true;
    }
    if (!seenDigit) {
      throw new IllegalArgumentException(field + " must be numeric");
    }
    return trimmed;
  }

  private static String normalizeOptional(final String value) {
    if (value == null) {
      return null;
    }
    final String trimmed = value.trim();
    return trimmed.isEmpty() ? null : trimmed;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/subscriptions/SubscriptionUsageRequest.kt
package org.hyperledger.iroha.sdk.subscriptions

import org.hyperledger.iroha.sdk.client.JsonEncoder

/** Request payload for subscription usage recording. */
class SubscriptionUsageRequest(
    authority: String,
    privateKey: String,
    unitKey: String,
    delta: String,
    usageTriggerId: String? = null,
) {
    val authority: String = requireNonBlank(authority, "authority")
    val privateKey: String = requireNonBlank(privateKey, "private_key")
    val unitKey: String = requireNonBlank(unitKey, "unit_key")
    val delta: String = requireNumericLiteral(delta, "delta")
    val usageTriggerId: String? = normalizeOptional(usageTriggerId)

    fun toJsonMap(): Map<String, Any> = buildMap {
        put("authority", authority)
        put("private_key", privateKey)
        put("unit_key", unitKey)
        put("delta", delta)
        this@SubscriptionUsageRequest.usageTriggerId?.let { put("usage_trigger_id", it) }
    }

    fun toJsonBytes(): ByteArray =
        JsonEncoder.encode(toJsonMap()).toByteArray(Charsets.UTF_8)
}

private fun requireNonBlank(value: String, field: String): String {
    val trimmed = value.trim()
    check(trimmed.isNotEmpty()) { "$field is required" }
    return trimmed
}

private fun requireNumericLiteral(value: String, field: String): String {
    val trimmed = requireNonBlank(value, field)
    var index = 0
    val first = trimmed[0]
    if (first == '+' || first == '-') {
        require(first != '-') { "$field must be non-negative" }
        index = 1
    }
    require(index < trimmed.length) { "$field must be numeric" }
    var seenDot = false
    var seenDigit = false
    for (i in index until trimmed.length) {
        val ch = trimmed[i]
        if (ch == '.') {
            require(!seenDot) { "$field must be numeric" }
            seenDot = true
            continue
        }
        require(ch in '0'..'9') { "$field must be numeric" }
        seenDigit = true
    }
    require(seenDigit) { "$field must be numeric" }
    return trimmed
}

private fun normalizeOptional(value: String?): String? {
    if (value == null) return null
    val trimmed = value.trim()
    return trimmed.ifEmpty { null }
}
